kernel/src/actions/schemas.rs

```diff
@@ -1,254 +1,44 @@
-            StructField::new("flavor", DataType::STRING, false),
-            tags_field(),
-        ]),
-        true,
-    );
-    // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#sidecar-file-information
-    static ref SIDECAR_FIELD: StructField = StructField::new(
-        "sidecar",
-        StructType::new(vec![
-            StructField::new("path", DataType::STRING, false),
-            StructField::new("sizeInBytes", DataType::LONG, false),
-            StructField::new("modificationTime", DataType::LONG, false),
-            StructField::new("type", DataType::STRING, false),
-            tags_field(),
-        ]),
-        true,
-    );
-
-    static ref LOG_SCHEMA: StructType = StructType::new(
-        vec![
-            ADD_FIELD.clone(),
-            CDC_FIELD.clone(),
-            COMMIT_INFO_FIELD.clone(),
-            DOMAIN_METADATA_FIELD.clone(),
-            METADATA_FIELD.clone(),
-            PROTOCOL_FIELD.clone(),
-            REMOVE_FIELD.clone(),
-            TXN_FIELD.clone(),
-        ]
-    );
-}
-
-fn tags_field() -> StructField {
-    StructField::new(
-        "tags",
-        MapType::new(DataType::STRING, DataType::STRING, true),
-        true,
-    )
-}
-
-fn partition_values_field() -> StructField {
-    StructField::new(
-        "partitionValues",
-        MapType::new(DataType::STRING, DataType::STRING, true),
-        false,
-    )
+/// A trait that allows getting a `StructField` based on the provided name and nullability
+pub(crate) trait GetField {
+    fn get_field(name: impl Into<String>) -> StructField;
 }
 
-fn deletion_vector_field() -> StructField {
-    StructField::new(
-        "deletionVector",
-        DataType::Struct(Box::new(StructType::new(vec![
-            StructField::new("storageType", DataType::STRING, false),
-            StructField::new("pathOrInlineDv", DataType::STRING, false),
-            StructField::new("offset", DataType::INTEGER, true),
-            StructField::new("sizeInBytes", DataType::INTEGER, false),
-            StructField::new("cardinality", DataType::LONG, false),
-        ]))),
-        true,
-    )
+macro_rules! impl_get_field {
+    ( $(($rust_type: ty, $kernel_type: expr)), * ) => {
+        $(
+            impl GetField for $rust_type {
+                fn get_field(name: impl Into<String>) -> StructField {
+                    StructField::new(name, $kernel_type, false)
+                }
+            }
+        )*
+    };
 }
 
-#[cfg(test)]
-pub(crate) fn log_schema() -> &'static StructType {
-    &LOG_SCHEMA
+impl_get_field!(
+    (String, DataType::STRING),
+    (i64, DataType::LONG),
+    (i32, DataType::INTEGER),
+    (i16, DataType::SHORT),
+    (char, DataType::BYTE),
+    (f32, DataType::FLOAT),
+    (f64, DataType::DOUBLE),
+    (bool, DataType::BOOLEAN),
+    (HashMap<String, String>, MapType::new(DataType::STRING, DataType::STRING, false)),
+    (HashMap<String, Option<String>>, MapType::new(DataType::STRING, DataType::STRING, true)),
```

Comment:
Do we have a ticket somewhere that tracks getting rid of this map-of-option-of-string thing?
(it keeps coming up in various PR)

Delta was originally written in Scala (Java variant), where there's no such thing as a nullable map entry: looking up a non-existent key returns null as a sentinel value. That makes it hard for me to imagine a case where map-of-option-of-string could ever be anything but a semantic overhead.

I guess we need to double check whether the json-serialized form of a string-string map somehow allows null entries. But even if it does somehow happen, I'd rather filter out such entries at parsing time rather than propagate them through the rest of the system?